src/main/java/seedu/address/model/person/PersonMatchesKeywordsPredicate.java

```diff
@@ -0,0 +1,100 @@
+package seedu.address.model.person;
+
+import seedu.address.commons.util.StringUtil;
+
+import java.util.function.Predicate;
+import java.util.Arrays;
+
+
+public class PersonMatchesKeywordsPredicate implements Predicate<Person> {
+    private final String keywords;
+
+    public PersonMatchesKeywordsPredicate(String keywords) {
+        this.keywords = keywords;
+    }
+
+    static int getLevenshteindist(String str1, String str2) {
+
+        int[][] arr = new int[str1.length() + 1][str2.length() + 1];
+
+        for (int i = 0; i <= str1.length(); i++)
+        {
+            for (int j = 0; j <= str2.length(); j++) {
+                if (i == 0) {
+                    arr[i][j] = j;
+                }
+
+                else if (j == 0) {
+                    arr[i][j] = i;
+                }
+
+                else {
+                    arr[i][j] = minEdits(arr[i - 1][j - 1]
+                                    + numOfReplacement(str1.charAt(i - 1),str2.charAt(j - 1)),
+                            arr[i - 1][j] + 1,
+                            arr[i][j - 1] + 1);
+                }
+            }
+        }
+        return arr[str1.length()][str2.length()];
+    }
+
+
+    public static double findSimilarity(String x, String y) {
+        double maxLength = Double.max(x.length(), y.length());
+        if (maxLength > 0) {
+            return (maxLength - getLevenshteindist(x, y)) / maxLength;
+        }
+        return 1.0;
+    }
+
+    static int numOfReplacement(char c1, char c2) {
+        return c1 == c2 ? 0 : 1;
+    }
+
+    static int minEdits(int... nums) {
+        return Arrays.stream(nums).min().orElse(Integer.MAX_VALUE);
+    }
+
+    private boolean matchesName(Person person) {
+        return findSimilarity(keywords,person.getName().fullName) > 0.5 ||
+                StringUtil.containsWordIgnoreCase(person.getName().fullName, keywords);
+    }
+
+    private boolean matchesAddress(Person person) {
+        return findSimilarity(keywords,person.getAddress().value) > 0.5 ||
+                StringUtil.containsWordIgnoreCase(person.getAddress().value, keywords);
+    }
+
+    private boolean matchesRole(Person person) {
+        return findSimilarity(keywords, String.valueOf(person.getRole().get())) > 0.5 ||
+                StringUtil.containsWordIgnoreCase(String.valueOf(person.getRole().get()), keywords);
```

Comment:
For Optional field, this piece of code might be cleaner.
```suggestion
        return person.getRole().ifPresent(r -> 
            findSimilarity(keywords, r.role) 
            || StringUtil.containsWordIgnoreCase(r.role, keywords))
            .orElse(false);
```